File: UnityGame/Assets/_Game/Career/CareerStatistics.cs
using System;
using System.Collections.Generic;

namespace CricketGame.Career
{
    [Serializable]
    public class BattingStatistics
    {
        public int matches;
        public int innings;
        public int notOuts;
        public int runs;
        public int highestScore;
        public int ballsFaced;
        public int fours;
        public int sixes;
        public int fifties;
        public int hundreds;

        public float Average
        {
            get { return innings - notOuts > 0 ? (float)runs / (innings - notOuts) : runs; }
        }

        public float StrikeRate
        {
            get { return ballsFaced > 0 ? (float)runs / ballsFaced * 100f : 0f; }
        }
    }

    [Serializable]
    public class BowlingStatistics
    {
        public int matches;
        public int innings;
        public float overs;
        public int maidens;
        public int runsConceded;
        public int wickets;
        public int bestWickets;
        public int bestRunsConceded;
        public int fiveWicketHauls;

        public float Economy
        {
            get { return overs > 0 ? runsConceded / overs : 0f; }
        }

        public float Average
        {
            get { return wickets > 0 ? (float)runsConceded / wickets : 0f; }
        }

        public float StrikeRate
        {
            get { return wickets > 0 ? (overs * 6f) / wickets : 0f; }
        }
    }

    [Serializable]
    public class FieldingStatistics
    {
        public int catches;
        public int runOuts;
        public int stumpings;
    }

    [Serializable]
    public class TournamentStatisticsRecord
    {
        public string tournamentId;
        public string tournamentName;
        public BattingStatistics batting = new BattingStatistics();
        public BowlingStatistics bowling = new BowlingStatistics();
        public FieldingStatistics fielding = new FieldingStatistics();
    }

    [Serializable]
    public class CareerStatistics
    {
        public BattingStatistics allTimeBatting = new BattingStatistics();
        public BowlingStatistics allTimeBowling = new BowlingStatistics();
        public FieldingStatistics allTimeFielding = new FieldingStatistics();
        public List<TournamentStatisticsRecord> byTournament = new List<TournamentStatisticsRecord>();

        public void RecordMatchBatting(int runsScored, int balls, int f4, int f6, bool isOut)
        {
            allTimeBatting.matches++;
            allTimeBatting.innings++;
            allTimeBatting.runs += runsScored;
            allTimeBatting.ballsFaced += balls;
            allTimeBatting.fours += f4;
            allTimeBatting.sixes += f6;
            if (!isOut) allTimeBatting.notOuts++;
            if (runsScored > allTimeBatting.highestScore) allTimeBatting.highestScore = runsScored;
            if (runsScored >= 100) allTimeBatting.hundreds++;
            else if (runsScored >= 50) allTimeBatting.fifties++;
        }

        public void RecordMatchBowling(float oversBowled, int maidensBowled, int runsGiven, int wkts)
        {
            allTimeBowling.matches++;
            allTimeBowling.innings++;
            allTimeBowling.overs += oversBowled;
            allTimeBowling.maidens += maidensBowled;
            allTimeBowling.runsConceded += runsGiven;
            allTimeBowling.wickets += wkts;
            if (wkts >= 5) allTimeBowling.fiveWicketHauls++;
        }

        public void RecordMatchFielding(int catchCount, int runOutCount, int stumpingCount)
        {
            allTimeFielding.catches += catchCount;
            allTimeFielding.runOuts += runOutCount;
            allTimeFielding.stumpings += stumpingCount;
        }
    }
}

File: src/player/CareerStatistics.ts
import { CareerLevel } from '../career/CareerLevel';

export interface BattingStats {
  matches: number;
  innings: number;
  notOuts: number;
  runs: number;
  ballsFaced: number;
  highestScore: number;
  isHighestScoreNotOut: boolean;
  average: number;
  strikeRate: number;
  fifties: number;
  hundreds: number;
  fours: number;
  sixes: number;
  ducks: number;
}

export interface BowlingStats {
  matches: number;
  overs: number;
  maidens: number;
  runsConceded: number;
  wickets: number;
  bestBowlingRuns: number;
  bestBowlingWickets: number;
  average: number;
  economy: number;
  threeWicketHauls: number;
  fiveWicketHauls: number;
}

export interface FieldingStats {
  catches: number;
  runOuts: number;
  stumpings: number;
}

export interface LevelStatisticsBreakdown {
  level: CareerLevel;
  batting: BattingStats;
  bowling: BowlingStats;
  fielding: FieldingStats;
}

export interface TournamentStatisticsRecord {
  tournamentId: string;
  tournamentName: string;
  level: CareerLevel;
  year: number;
  batting: BattingStats;
  bowling: BowlingStats;
  fielding: FieldingStats;
}

export interface CareerStatisticsContainer {
  allTime: {
    batting: BattingStats;
    bowling: BowlingStats;
    fielding: FieldingStats;
  };
  byLevel: Record<string, LevelStatisticsBreakdown>;
  byTournament: TournamentStatisticsRecord[];
}

export class CareerStatisticsTracker {
  public static createEmptyBattingStats(): BattingStats {
    return {
      matches: 0,
      innings: 0,
      notOuts: 0,
      runs: 0,
      ballsFaced: 0,
      highestScore: 0,
      isHighestScoreNotOut: false,
      average: 0.0,
      strikeRate: 0.0,
      fifties: 0,
      hundreds: 0,
      fours: 0,
      sixes: 0,
      ducks: 0
    };
  }

  public static createEmptyBowlingStats(): BowlingStats {
    return {
      matches: 0,
      overs: 0,
      maidens: 0,
      runsConceded: 0,
      wickets: 0,
      bestBowlingRuns: 0,
      bestBowlingWickets: 0,
      average: 0.0,
      economy: 0.0,
      threeWicketHauls: 0,
      fiveWicketHauls: 0
    };
  }

  public static createEmptyFieldingStats(): FieldingStats {
    return {
      catches: 0,
      runOuts: 0,
      stumpings: 0
    };
  }

  public static createInitialContainer(): CareerStatisticsContainer {
    return {
      allTime: {
        batting: this.createEmptyBattingStats(),
        bowling: this.createEmptyBowlingStats(),
        fielding: this.createEmptyFieldingStats()
      },
      byLevel: {},
      byTournament: []
    };
  }

  /**
   * Applies match performance to all-time, level-specific, and tournament-specific stats.
   */
  public static recordMatch(
    container: CareerStatisticsContainer,
    level: CareerLevel,
    tournamentId: string,
    tournamentName: string,
    perf: {
      runs: number;
      balls: number;
      fours: number;
      sixes: number;
      isOut: boolean;
      didBat: boolean;
      overs: number;
      maidens: number;
      runsConceded: number;
      wickets: number;
      didBowl: boolean;
      catches: number;
      runOuts: number;
      stumpings: number;
    }
  ): void {
    // 1. Ensure Level Breakdown exists
    if (!container.byLevel[level]) {
      container.byLevel[level] = {
        level,
        batting: this.createEmptyBattingStats(),
        bowling: this.createEmptyBowlingStats(),
        fielding: this.createEmptyFieldingStats()
      };
    }
    const levelStats = container.byLevel[level];

    // 2. Ensure Tournament Record exists
    let tourneyRecord = container.byTournament.find(t => t.tournamentId === tournamentId);
    if (!tourneyRecord) {
      tourneyRecord = {
        tournamentId,
        tournamentName,
        level,
        year: new Date().getFullYear(),
        batting: this.createEmptyBattingStats(),
        bowling: this.createEmptyBowlingStats(),
        fielding: this.createEmptyFieldingStats()
      };
      container.byTournament.push(tourneyRecord);
    }

    // Helper to update Batting Stats
    const updateBatting = (b: BattingStats) => {
      b.matches++;
      if (perf.didBat) {
        b.innings++;
        b.runs += perf.runs;
        b.ballsFaced += perf.balls;
        b.fours += perf.fours;
        b.sixes += perf.sixes;

        if (!perf.isOut) {
          b.notOuts++;
        } else if (perf.runs === 0) {
          b.ducks++;
        }

        if (perf.runs > b.highestScore || (perf.runs === b.highestScore && !perf.isOut)) {
          b.highestScore = perf.runs;
          b.isHighestScoreNotOut = !perf.isOut;
        }

        if (perf.runs >= 100) {
          b.hundreds++;
        } else if (perf.runs >= 50) {
          b.fifties++;
        }

        const dismissals = b.innings - b.notOuts;
        b.average = dismissals > 0 ? Math.round((b.runs / dismissals) * 100) / 100 : b.runs;
        b.strikeRate = b.ballsFaced > 0 ? Math.round((b.runs / b.ballsFaced) * 10000) / 100 : 0.0;
      }
    };

    // Helper to update Bowling Stats
    const updateBowling = (bw: BowlingStats) => {
      bw.matches++;
      if (perf.didBowl) {
        bw.overs += perf.overs;
        bw.maidens += perf.maidens;
        bw.runsConceded += perf.runsConceded;
        bw.wickets += perf.wickets;

        if (perf.wickets >= 5) {
          bw.fiveWicketHauls++;
        } else if (perf.wickets >= 3) {
          bw.threeWicketHauls++;
        }

        // Check best bowling
        if (
          perf.wickets > bw.bestBowlingWickets ||
          (perf.wickets === bw.bestBowlingWickets && perf.runsConceded < bw.bestBowlingRuns) ||
          bw.bestBowlingWickets === 0
        ) {
          bw.bestBowlingWickets = perf.wickets;
          bw.bestBowlingRuns = perf.runsConceded;
        }

        bw.average = bw.wickets > 0 ? Math.round((bw.runsConceded / bw.wickets) * 100) / 100 : 0.0;
        bw.economy = bw.overs > 0 ? Math.round((bw.runsConceded / bw.overs) * 100) / 100 : 0.0;
      }
    };

    // Helper to update Fielding Stats
    const updateFielding = (f: FieldingStats) => {
      f.catches += perf.catches;
      f.runOuts += perf.runOuts;
      f.stumpings += perf.stumpings;
    };

    // Apply to all 3 scopes
    updateBatting(container.allTime.batting);
    updateBowling(container.allTime.bowling);
    updateFielding(container.allTime.fielding);

    updateBatting(levelStats.batting);
    updateBowling(levelStats.bowling);
    updateFielding(levelStats.fielding);

    updateBatting(tourneyRecord.batting);
    updateBowling(tourneyRecord.bowling);
    updateFielding(tourneyRecord.fielding);
  }
}
